### PR TITLE
Added the grammar support for type extensions

### PR DESCRIPTION
**Pull request by bbakerman**
see #906

This adds the Antlr support for parsing an AST of type extensions

### CODE REVIEW

File: src/main/antlr/Graphql.g4 [hunk 1]
```diff
@@ -1,5 +1,5 @@
 grammar Graphql;
-import GraphqlSDL, GraphqlOperation;
```
Comment (bbakerman) [inline src/main/antlr/Graphql.g4]: this is needed more for IDEA and its "run ANLTR rule" support but it does no harm

File: src/main/java/graphql/parser/GraphqlAntlrToLanguage.java [hunk 2]
```diff
@@ -636,17 +631,45 @@ public Void visitInputObjectTypeDefinition(GraphqlParser.InputObjectTypeDefiniti
...
+        newNode((AbstractNode) def, ctx);
         result.getDefinitions().add(def);
-        addContextProperty(ContextProperty.TypeExtensionDefinition, def);
+        addContextProperty(ctxProperty, def);
         super.visitChildren(ctx);
         popContext();
         return null;
     }
 
+    @Override
```
Comment (bbakerman) [inline src/main/java/graphql/parser/GraphqlAntlrToLanguage.java]: New naming in the .g4 generates new entry points.  But I think the name change is a little better

File: src/main/antlr/Graphql.g4 (hunk 1 above)
Comment (andimarek) [reply src/main/antlr/Graphql.g4]: I don't understand: we need to import the rules we wanna use ... why is this only needed for IDEA?

File: src/main/java/graphql/parser/GraphqlAntlrToLanguage.java (hunk 2 above)
Comment (andimarek) [reply src/main/java/graphql/parser/GraphqlAntlrToLanguage.java]: 👍

File: src/test/groovy/graphql/parser/IDLParserTest.groovy [hunk 3]
```diff
@@ -582,5 +588,162 @@ input Gun {
...
+
+        when:
+        def doc = new Parser().parseDocument(input)
+
+        then:
+
+        // object type extension
+        fromDoc(doc, 0, ObjectTypeDefinition).name == "Query"
+
+        fromDoc(doc, 1, TypeExtensionDefinition).name == "Query"
```
Comment (andimarek) [inline src/test/groovy/graphql/parser/IDLParserTest.groovy]: I think we should use this PR and rename TypeExtensionDefinition to ObjectTypeExtension to make it in line with the other names

File: src/test/groovy/graphql/parser/IDLParserTest.groovy [hunk 4]
```diff
@@ -582,5 +588,162 @@ input Gun {
         (typeDef2.getImplements()[1] as TypeName).getName() == 'Baz'
     }
 
+    def "basic type extensions"() {
```
Comment (andimarek) [inline src/test/groovy/graphql/parser/IDLParserTest.groovy]: how about splitting it up in several tests?

File: src/main/antlr/Graphql.g4 (hunk 1 above)
Comment (bbakerman) [reply src/main/antlr/Graphql.g4]: IDEA needs this to be able to use the "Run Rule" plugin that helps you run the grammar interactively.  Without this you can do that.  It does not work out the transitive dependency

However at actual generation time, it works out the transitive dependency ok.

So its harmless but really helps in development that you can run the rule interactively in IDEA.

Check it out.  Its a cool IDEA feature

image

File: src/test/groovy/graphql/parser/IDLParserTest.groovy (hunk 4 above)
Comment (bbakerman) [reply src/test/groovy/graphql/parser/IDLParserTest.groovy]: ok'

File: src/test/groovy/graphql/parser/IDLParserTest.groovy (hunk 3 above)
Comment (bbakerman) [reply src/test/groovy/graphql/parser/IDLParserTest.groovy]: ok

**Comment by bbakerman** [comment]
OK PR feedback in - also added AST printing

File: src/main/antlr/Graphql.g4 (hunk 1 above)
Comment (andimarek) [reply src/main/antlr/Graphql.g4]: ah ok .. the only consequence is that we actually generate code for every rule in graphqlCommon, even if it is not really used. But on the other side there should be not unused rule, so 👍

**Comment by andimarek** [comment]
breaking change because we rename TypeExtensionDefinition to ObjectTypeExtensionDefinition

**Pull request merged**